app/controllers/MessageResponseController.scala

```diff
@@ -0,0 +1,88 @@
+import models.MessageResponse
+import models.MessageSender
+import models.MessageType
+import models.UnloadingPermissionResponse
+import play.api.Logger
+import play.api.mvc.Action
+import play.api.mvc.ControllerComponents
+import repositories.ArrivalMovementRepository
+import services.ArrivalMovementService
+import services.XmlValidationService
+import uk.gov.hmrc.play.bootstrap.controller.BackendController
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.Failure
+import scala.util.Success
+import scala.xml.NodeSeq
+
+class MessageResponseController @Inject()(cc: ControllerComponents,
+                                          arrivalMovementService: ArrivalMovementService,
+                                          getArrival: GetArrivalForWriteActionProvider,
+                                          arrivalMovementRepository: ArrivalMovementRepository,
+                                          xmlValidationService: XmlValidationService)(implicit ec: ExecutionContext)
+    extends BackendController(cc) {
+
+  private val logger = Logger(getClass)
+
+  def post(messageSender: MessageSender): Action[NodeSeq] = getArrival(messageSender.arrivalId)(parse.xml).async {
+    implicit request =>
+      val xml: NodeSeq = request.request.body
+
+      val messageResponse: Option[MessageResponse] = request.headers.get("X-Message-Type") match {
+        case Some(MessageType.GoodsReleased.code)       => Some(GoodsReleasedResponse)
+        case Some(MessageType.UnloadingPermission.code) => Some(UnloadingPermissionResponse)
+        case invalidResponse =>
+          logger.error(s"Received the following invalid response for X-Message-Type: $invalidResponse")
+          None
+      }
+
+      messageResponse match {
+        case Some(response) =>
+          xmlValidationService.validate(xml.toString, response.xsdFile) match {
+            case Success(_) =>
+              arrivalMovementService.makeMessage(messageSender.messageCorrelationId, response.messageType)(xml) match {
+                case Some(message) =>
+                  val newState: ArrivalStatus = request.arrival.status.transition(response.messageReceived)
+                  arrivalMovementRepository.addResponseMessage(request.arrival.arrivalId, message, newState).map {
+                    case Success(_) => Ok
+                    case Failure(e) =>
+                      logger.error(s"Failure to add message to movement. Exception: ${e.getMessage}")
+                      InternalServerError
+                  }
+                case None =>
+                  logger.error(s"Failure to parse message")
+                  Future.successful(BadRequest)
+              }
+            case Failure(e) =>
+              logger.error(s"Failure to validate against XSD. Exception: ${e.getMessage}")
+              Future.successful(BadRequest)
+          }
+        case None =>
+          Future.successful(BadRequest)
```

Comment:
We need a named logger here since we're rejecting the incoming message

Reply:
Does the logger on line 57 - 59 not cover this. We could move it down to this layer if it makes more sense.